pkg/cluster/denyassignment.go

```diff
@@ -0,0 +1,49 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+	"github.com/Azure/ARO-RP/pkg/util/deployment"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+)
+
+func (m *manager) createOrUpdateDenyAssignment(ctx context.Context) error {
+	if m.env.DeploymentMode() != deployment.Production {
+		// only need this upgrade in production, where there are DenyAssignments
+		return nil
+	}
+
+	resourceGroup := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+
+	clusterSPObjectID, err := m.aad.GetServicePrincipalID(ctx)
+	if err != nil {
+		return err
+	}
+
+	denyAssignments, err := m.denyassignments.ListForResourceGroup(ctx, resourceGroup, "")
+	if err != nil {
+		return err
+	}
+
+	for _, assignment := range denyAssignments {
+		if assignment.DenyAssignmentProperties.ExcludePrincipals != nil {
+			for _, ps := range *assignment.DenyAssignmentProperties.ExcludePrincipals {
+				if *ps.ID == clusterSPObjectID {
+					return nil
+				}
+			}
+		}
+	}
+
+	t := &arm.Template{
+		Schema:         "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+		ContentVersion: "1.0.0.0",
+		Resources:      []*arm.Resource{m.denyAssignments(clusterSPObjectID)},
+	}
+
+	return m.arm.Deploy(ctx, resourceGroup, "storage", t, nil)
```

Comment:
I suspect we're going to reach the end of the road with this strategy and will need to have a client to call.  Maybe we're lucky and it exists in the Go Azure SDK now?